client/verta/verta/_internal_utils/_utils.py

```diff
@@ -234,32 +234,128 @@ def __len__(self):
+
+            # cast operator into protobuf enum variant
+            operator = self._OP_MAP[operator]
+
+            try:
+                value = float(value)
+            except ValueError:  # not a number
+                # parse value
+                try:
+                    expr_node = ast.parse(value, mode='eval')
+                except SyntaxError:
+                    e = ValueError("value `{}` must be a number or string literal".format(value))
+                    six.raise_from(e, None)
+                value_node = expr_node.body
+                if type(value_node) is ast.Str:
+                    value = value_node.s
+                elif type(value_node) is ast.Compare:
+                    e = ValueError("predicate `{}` must be a two-operand comparison".format(predicate))
+                    six.raise_from(e, None)
+                else:
+                    e = ValueError("value `{}` must be a number or string literal".format(value))
+                    six.raise_from(e, None)
+
+            new_list._msg.predicates.append(  # pylint: disable=no-member
+                _CommonService.KeyValueQuery(
+                    key=key, value=_utils.python_to_val_proto(value),
+                    operator=operator,
+                )
             )
-        raise_for_http_error(response)
 
-        response_msg = json_to_proto(body_to_json(response), msg.Response)
-        return response_msg
+        return new_list
 
-    def _get_records(self, response_msg):
-        """Get the attribute of `response_msg` that is not `total_records`."""
+    def sort(self, key, descending=False):
+        """
+        Sorts the results from this collection by `key`.
+
+        A `key` is a string containing a dot-delimited property such as
+        ``metrics.accuracy``.
+
+        Parameters
+        ----------
+        key : str
+            Dot-delimited property.
+        descending : bool, default False
+            Order in which to return sorted results.
+
+        Returns
+        -------
+        The same type of object given in the input.
+
+        Examples
+        --------
+        .. code-block:: python
+
+            runs.sort("metrics.accuracy")
+            # <ExperimentRuns containing 3 runs>
+
+        """
+        if ret_all_info:
+            warnings.warn("`ret_all_info` is deprecated and will removed in a later version",
+                          category=FutureWarning)
+
+        if key.split('.')[0] not in self._VALID_QUERY_KEYS:
+            raise ValueError("key `{}` is not a valid key for querying;"
+                             " currently supported keys are: {}".format(key, self._VALID_QUERY_KEYS))
+
+        new_list = copy.deepcopy(self)
+
+        new_list._msg.sort_key = key
+        new_list._msg.ascending = not descending
+
+        return new_list
+
+    def _call_back_end(self, msg):
+        """Find the request in the backend and returns (elements, total count)."""
         raise NotImplementedError
```

Comment:
The [subclasses for `Dataset` and `DatasetVersion`](_call_back_end) probably also need to implement this.
I could do that in a separate PR, if this should be merged in first.

For my reference, [`Dataset`'s](https://github.com/VertaAI/modeldb/blob/c882224/client/verta/verta/client.py#L684) is used here, [`DatasetVersion`'s](https://github.com/VertaAI/modeldb/blob/c882224/client/verta/verta/_dataset.py#L196) is used here.

Reply:
Yeah, it can be done after. Also, be mindful about not manipulating internal fields of other objects. Use things like `with_*` instead to provide API between objects.

Reply:
I'd also recommend sticking with the API of getting the full list by default and then filtering. It will make things easier to build docs.